docs/book/v1/async-tasks.md

```diff
@@ -0,0 +1,129 @@
+
+### Configuring the Server Process
+
+In order to take advantage of this feature within a Zend Expressive application running in the Swoole environment, you
+will first need to configure the server to start up task workers. In your local configuration for the server, you'll
+ need to add `task_worker_num` 
+
+```php
+'zend-expressive-swoole' => [
+    'swoole-http-server' => [
+        'host' => '127.0.0.1',
+        'port' => 8080,
+        'options' => [
+            'worker_num' => 4, // The number of HTTP Server Workers
+            'task_worker_num' => 4, // The number of Task Workers
+        ],
+    ],
+];
+```
+
+### Handling Task Events and Doing the Work
+
+The Swoole server will now require that 2 new event callbacks are non null. These are the `onTask` and `onFinish`
+events. Without these setup, the server will refuse to start.
+
+In order to simplify this example, the work to be done, and the event callbacks will be concise functions designed to
+illustrate the logic flow:
+
+```php
+use Swoole\Http\Server as HttpServer;
+
+$actualWork = function () {
+    sleep(5);
+    return 'This will be available in the onTask callback';
+});
+
+$serverOnTaskCallback = function (HttpServer $server, int $taskId, int $sourceWorkerId, $dataForWorker) use ($actualWork) {
+    // Do the work, syncronously
+    // $dataForWorker can be any value except for a resource, it is received by triggering a task from the
+    // Swoole server process like this: $server->task($data)
+    $result = $actualWork();
+    return 'Actual Work is Done';
+});
+
+$serverOnFinishCallback = function (HttpServer $server, int $taskId, $userData) {
+    // Perhaps log completion of the work.
+    // $userData === 'Actual Work is Done';
+};
+```
+
+### Providing Task Event Handlers to the HTTP Server Instance
+
+Now the work is defined, we need to provide these callbacks to the HTTP server. This is best achieved by utilising a
+delegator factory to decorate the the server instance at the point of its construction.
+
+In Zend Expressive Swoole, the Http server factory is aliased to Swoole's FQCN, `Swoole\Http\Server`, so a delegator
+factory might look something like this:
+
+```php
+<?php
+// src/App/SwooleHttpDelegatorFactory.php
+namespace App;
+
+use Psr\Container\ContainerInterface;
+use Swoole\Http\Server as SwooleServer;
+
+class SwooleHttpDelegatorFactory
+{
+    public function __invoke(ContainerInterface $container, $serviceName, callable $callback) : SwooleServer
+     {
+        $server = $callback();
+        // We're pretending that the callbacks exist in this scope.
+        // Typically, you might retrieve a service from the container and inject the server instance into that service
+        // in order to configure your tasks, or use the service in a callback directly here.
+        $server->on('Task', $serverOnTaskCallback);
+        $server->on('Finish', $serverOnFinishCallback);
+        return $server;
+    }
+}
+```
```

Comment:
The previous section and the above confused me a bit, particularly after [reading the Swoole project async task docs](https://github.com/swoole/swoole-docs/blob/31ca00158567377b168151f9ca186c35307cfe06/get-started/examples/async_task.md). My understanding is that in order to enable tasks, you need to do exactly one thing: register a listener on the 'task' event. Registering a 'finish' event is optional, per that document, and only necessary if your 'task' listener returns a value; you can also pass a callback to `task()` that will execute on task completion.

The `task` listener itself figures out what to do with the data, and could potentially perform actions based on both the data passed to it, as well as the result of processing, which means we can make the example more succinct and opinionated.

As an example, we could do something that composes both a logger and a message notifier, and it might look something like this:

```php
use Psr\EventDispatcher\MessageInterface;
use Psr\EventDispatcher\MessageNotifierInterface;
use Psr\Log\LoggerInterface;

class TaskWorker
{
    private $notifier;
    private $logger;

    public function __construct(LoggerInterface $logger, MessageNotifierInterface $notifier)
    {
        $this->logger = $logger;
        $this->notifier = $notifier;
    }

    public function __invoke($server, $taskId, $fromId, $data)
    {
        if (! $data instanceof MessageInterface) {
            $this->logger->error('Invalid data provided to task worker: {type}', ['type' => is_object($data) ? get_class($data) : gettype($data)]);
            return;
        }
        $this->logger->info('Starting work on task {taskId} using data: {data}', ['taskId' => $taskId, 'data' => json_encode($data)]);
        try {
            $this->notifier->notify($data);
        } catch (\Throwable $e) {
            $this->logger->error('Error processing task {taskId}: {error}', ['taskId' => $taskId, 'error' => $e->getTraceAsString()]);
        }
    }
}

$server->on('task', $container->get(TaskWorker::class));
```

This makes it more clear, at least to me, what you should be passing to `on('task')`, and what that callback's responsibilities are.

One thing I'm unclear about is whether Swoole allows registering multiple task listeners. If it doesn't, we need to document that, so folks don't try registering multiple times. If it does, what's the behavior? Does each get the same `$data`? What gets passed to `finish` in that situation? We should document those things here.

Reply:
I haven't finished reading this just yet - but `on('Finish')` is required in my experience - I get a fatal error anytime I omit that handler. I'm pretty confident that swoole does not allow multiple listeners - I've tested this with some events, but I'll go away and check the same holds true for task listeners etc.

Reply:
`$server->task()` signature is `$server->task(mixed $userData, int $destinationWorkerId = -1)` - The only way of processing tasks is in `$server->on('Task', $callable)` - Just been testing this out and reading up here: https://wiki.swoole.com/wiki/page/134.html